### PR TITLE
QR Code Based Authentication

### PR DESCRIPTION
**Pull request by Isurika1998**
## Purpose
This PR is to add the QR Code Based Authentication.

Related Issue : https://github.com/wso2/product-is/issues/12833

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/QRAuthContextManager.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * This interface manages authentication context operations for QR based authentication.
+ */
+public interface QRAuthContextManager {
+
+    /**
+     * Store authentication context.
+     *
+     * @param key     Unique key for identifying the authentication context for the session
+     * @param context Authentication context
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/QRAuthContextManager.java:33]: Add (.) at the end of each comment. Check in other places as well,

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/QRJWTValidator.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+     * @return Boolean value for validation.
+     */
+    public JWTClaimsSet getValidatedClaimSet(String jwt, String publicKey) throws IdentityQRAuthException {
+
+        if (!isJWT(jwt)) {
+            throw new IdentityQRAuthException("Token is not a valid JWT.");
+        }
+
+        try {
+            SignedJWT signedJWT = SignedJWT.parse(jwt);
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/QRJWTValidator.java:65]: Can't this signedJWT be null?

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/QRJWTValidator.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+     * @param signedJWT    Signed JWT.
+     * @return Boolean value for signature validation.
+     * @throws IdentityQRAuthException Error when validating the signature.
+     */
+    private boolean validateSignature(String publicKeyStr, SignedJWT signedJWT) throws IdentityQRAuthException {
+
+        try {
+            byte[] publicKeyData = Base64.getDecoder().decode(publicKeyStr);
+            X509EncodedKeySpec spec = new X509EncodedKeySpec(publicKeyData);
+            KeyFactory kf = KeyFactory.getInstance(QRAuthCommonConstants.SIGNING_ALGORITHM);
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/QRJWTValidator.java:101]:
```suggestion
            KeyFactory keyFactory = KeyFactory.getInstance(QRAuthCommonConstants.SIGNING_ALGORITHM);
```

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/QRJWTValidator.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            long currentTimeInMillis = System.currentTimeMillis();
+            return currentTimeInMillis + timeStampSkewMillis >= notBeforeTimeMillis;
+        }
+        return false;
+    }
+
+    /**
+     * Validate legitimacy of JWT.
+     *
+     * @param tokenIdentifier JWT string
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/QRJWTValidator.java:145]: @return needs o be here

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/QRAuthCommonConstants.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+package org.wso2.carbon.identity.application.authenticator.qrcode.common;
+
+/**
+ * Constants for QR authentication common component.
+ */
+public class QRAuthCommonConstants {
+
+    public static final String SIGNING_ALGORITHM = "RSA";
+}
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/QRAuthCommonConstants.java:28]: Do we need this class

Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/QRAuthCommonConstants.java:28]: Do we need this class

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/impl/QRAuthContextManagerImpl.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+        QRAuthContextCache.getInstance().addToCacheByRequestId(
+                new QRAuthContextCacheKey(key), new QRAuthContextCacheEntry(context));
+    }
+
+    @Override
+    public AuthenticationContext getContext(String key) {
+
+        QRAuthContextCacheKey cacheKey = new QRAuthContextCacheKey(key);
+        return QRAuthContextCache.getInstance().getValueFromCacheByRequestId(cacheKey).getAuthenticationContext();
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/impl/QRAuthContextManagerImpl.java:44]: Can `QRAuthContextCache.getInstance().getValueFromCacheByRequestId(cacheKey)` this be null? if so let's do a null check

Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/impl/QRAuthContextManagerImpl.java:44]: Can `QRAuthContextCache.getInstance().getValueFromCacheByRequestId(cacheKey)` this be null? if so let's do a null check

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/pom.xml [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+                            org.apache.commons.logging;version="${org.apache.commons.version.range}",
+                            org.eclipse.equinox.http.helper,
+                            org.osgi.service.component;version="${org.osgi.service.version.range}",
+                            org.osgi.service.http;version="${org.osgi.service.version.range}",
+                            org.wso2.carbon.context.*; version="${carbon.kernel.package.import.version.range}",
+                            org.wso2.carbon.identity.application.authenticator.qrcode.dto,
+                            org.wso2.carbon.identity.application.authenticator.qrcode.common,
+                            org.wso2.carbon.identity.application.authenticator.qrcode.common.impl,
+                            org.wso2.carbon.identity.application.authenticator.qrcode.common.exception,
+                            org.wso2.carbon.user.core.*; version="${carbon.kernel.package.import.version.range}",
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/pom.xml:160]: let's avoid using "*".

Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/pom.xml:160]: let's avoid using "*".

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/src/main/java/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/QRServletConstants.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    public static final String AUTHENTICATED_USER = "authUser";
+    public static final String AUTH_DATA = "authData";
+    public static final String AUTH_STATUS = "authStatus";
+    public static final String AUTH_RESPONSE = "authResponse";
+    public static final String MEDIA_TYPE_JSON = "application/json";
+
+    /**
+     * Object holding authentication mobile response status.
+     */
+    public enum Status {
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/src/main/java/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/QRServletConstants.java:40]: Add new line

Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/src/main/java/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/QRServletConstants.java:40]: Add new line

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/src/main/java/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/QRServletConstants.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        COMPLETED, PENDING
+    }
+
+    /**
+     * Enum which contains error codes and corresponding error messages.
+     */
+    public enum ErrorMessages {
+
+        ERROR_CODE_AUTH_RESPONSE_TOKEN_NOT_FOUND(
+                "PBA-15001",
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/src/main/java/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/QRServletConstants.java:50]: What is this "PBA" stands for? Better to use an abbreviation related to your component

Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/src/main/java/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/QRServletConstants.java:50]: What is this "PBA" stands for? Better to use an abbreviation related to your component

Comment (Isurika1998) [reply components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/src/main/java/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/QRServletConstants.java]: Changed to "QBA" - QR Based Authenticator

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/impl/QRAuthContextManagerImpl.java (hunk 6 above)
Comment (Isurika1998) [reply components/org.wso2.carbon.identity.application.authenticator.qrcode.common/src/main/java/org/wso2/carbon/identity/application/authenticator/qrcode/common/impl/QRAuthContextManagerImpl.java:44]: null check is done inside `getValueFromCacheByRequestId(cacheKey)` method

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/pom.xml [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+                            org.wso2.carbon.context; version="${carbon.kernel.package.import.version.range}",
+                            org.wso2.carbon.identity.application.authenticator.qrcode.common,
+                            org.wso2.carbon.identity.application.authenticator.qrcode.common.impl,
+                            org.wso2.carbon.user.core.util; version="${carbon.kernel.package.import.version.range}",
+                            com.nimbusds.jose;version="${com.nimbusds.version.range}",
+                            com.nimbusds.jwt;version="${com.nimbusds.version.range}",
+                            org.wso2.carbon.identity.application.authentication.framework.inbound;version="${org.wso2.carbon.identity.version.range}",
+                            org.wso2.carbon.identity.application.authenticator.qrcode.common.exception,
+                        </Import-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/pom.xml:159]: avoid using DynamicImport-Package

File: components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/src/main/java/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/internal/QREndpointServiceComponent.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+            unbind = "unsetHttpService")
+    protected void setHttpService(HttpService httpService) {
+
+        this.httpService = httpService;
+    }
+
+    protected void unsetHttpService(HttpService httpService) {
+
+        this.httpService = null;
+    }
```
Comment (GANGANI) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/src/main/java/org.wso2.carbon.identity.application.authenticator.qrcode.servlet/internal/QREndpointServiceComponent.java:97]: Remove this if this is not used

File: components/org.wso2.carbon.identity.application.authenticator.qrcode/src/test/java/org.wso2.carbon.identity.application.authenticator.qrcode/QRAuthenticatorTest.java [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+    @Mock
+    private ServiceURL serviceURL;
+
+    @BeforeMethod
+    public void setUp() {
+        qrAuthenticator = new QRAuthenticator();
+        initMocks(this);
+    }
+
+//    @Test(description = "Test case for canHandle() method true case.")
```
Comment (ashensw) [inline components/org.wso2.carbon.identity.application.authenticator.qrcode/src/test/java/org.wso2.carbon.identity.application.authenticator.qrcode/QRAuthenticatorTest.java:96]: Any reason for commenting tests?

**Review by ashensw — APPROVED** [review]

**Review by GANGANI — APPROVED** [review]

**Pull request merged**